### PR TITLE
MONGOID-4977 (backport): Use recent driver APIs in Mongoid QueryCache for compatibility with sessions & modern retryable reads

### PR DESCRIPTION
**Pull request by egiurleo**
You can see the tests running in this patch: [https://spruce.mongodb.com/version/5f91d81c0ae606431f650bc7/tasks](https://spruce.mongodb.com/version/5f91d81c0ae606431f650bc7/tasks)

### CODE REVIEW

File: lib/mongoid/query_cache.rb [hunk 1]
```diff
@@ -224,34 +223,39 @@ module View
...
+              session = client.send(:get_session, @options)
+              read_with_retry(session, server_selector) do |server|
+                result = send_initial_query(server, session)
+                @cursor = cursor(result, server, session)
+              end
+            else
+              read_with_retry do
+                server = server_selector.select_server(cluster)
+                result = send_initial_query(server)
+                @cursor = cursor(result, server)
```
Comment (egiurleo) [inline lib/mongoid/query_cache.rb:242]: I do not love this code 😆 If you have suggestions about how to improve it, I am open to hearing them.

File: spec/mongoid/query_cache_spec.rb [hunk 2]
```diff
@@ -543,7 +557,7 @@
       Mongoid::QueryCache.enabled = true
       10.times { Band.create! }
 
-      Band.batch_size(4).all.any?
+      Band.batch_size(4).all
```
Comment (p-mongo) [inline spec/mongoid/query_cache_spec.rb:560]: This changes from executing the query to not executing, no?

File: lib/mongoid/query_cache.rb (hunk 1 above)
Comment (p-mongo) [reply lib/mongoid/query_cache.rb:242]: Can the method be called something other than `cursor` to avoid future confusion? Maybe `get_cursor` even?

Comment (egiurleo) [reply lib/mongoid/query_cache.rb:242]: Sure thing.

File: spec/mongoid/query_cache_spec.rb (hunk 2 above)
Comment (egiurleo) [reply spec/mongoid/query_cache_spec.rb:560]: Yup good catch. Meant to change it to `to_a`.

Comment (egiurleo) [reply spec/mongoid/query_cache_spec.rb:560]: Block methods like `any?` currently cause session leaks as described by [RUBY-2391](https://jira.mongodb.org/browse/RUBY-2391).

**Comment by egiurleo** [comment]
You can see the tests running in [patch 303](https://spruce.mongodb.com/version/5f9305412fbabe0fe81dd6a6/tasks).

**Review by p-mongo — APPROVED** [review]

**Pull request merged**